Opening file causes exception

This file, created with an earlier version of OR, causes an exception when loading into a recent snapshot build.

[Estes QCC Explorer (3221).zip](https://github.com/user-attachments/files/25186232/Estes.QCC.Explorer.3221.zip)

```
---------- Exception stack trace ----------
info.openrocket.core.util.BugException: BUG: Unknown error when opening file
at info.openrocket.swing.gui.main.BasicFrame.open(BasicFrame.java:1617)
at info.openrocket.swing.gui.main.BasicFrame.open(BasicFrame.java:1571)
at info.openrocket.swing.gui.main.BasicFrame.openAction(BasicFrame.java:1471)
at info.openrocket.swing.gui.main.BasicFrame.openAction(BasicFrame.java:1482)
at info.openrocket.swing.gui.main.BasicFrame$35.actionPerformed(BasicFrame.java:1113)
at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1972)
at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2313)
at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:374)
at java.desktop/com.apple.laf.ScreenMenuItem.actionPerformed(ScreenMenuItem.java:129)
at java.desktop/java.awt.MenuItem.processActionEvent(MenuItem.java:692)
at java.desktop/java.awt.MenuItem.processEvent(MenuItem.java:651)
at java.desktop/java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:379)
at java.desktop/java.awt.MenuComponent.dispatchEvent(MenuComponent.java:367)
at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:747)
at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:744)
at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Component: Pod Set not currently compatible with component: Stage
at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
at java.desktop/javax.swing.SwingWorker.get(SwingWorker.java:613)
at info.openrocket.swing.gui.main.BasicFrame.open(BasicFrame.java:1600)
... 31 more
Caused by: java.lang.IllegalStateException: Component: Pod Set not currently compatible with component: Stage
at info.openrocket.core.rocketcomponent.RocketComponent.addChild(RocketComponent.java:1914)
at info.openrocket.core.rocketcomponent.RocketComponent.addChild(RocketComponent.java:1868)
at info.openrocket.core.rocketcomponent.RocketComponent.addChild(RocketComponent.java:1882)
at info.openrocket.core.file.openrocket.importt.ComponentHandler.openElement(ComponentHandler.java:51)
at info.openrocket.core.file.simplesax.DelegatorHandler.startElement(DelegatorHandler.java:52)
at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:518)
at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:374)
at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)
at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
at info.openrocket.core.file.simplesax.SimpleSAX.readXML(SimpleSAX.java:56)
at info.openrocket.core.file.openrocket.importt.OpenRocketLoader.loadFromStream(OpenRocketLoader.java:49)
at info.openrocket.core.file.AbstractRocketLoader.load(AbstractRocketLoader.java:20)
at info.openrocket.core.file.GeneralRocketLoader.loadUsing(GeneralRocketLoader.java:256)
at info.openrocket.core.file.GeneralRocketLoader.loadRocket(GeneralRocketLoader.java:209)
at info.openrocket.core.file.GeneralRocketLoader.loadStep1(GeneralRocketLoader.java:170)
at info.openrocket.core.file.GeneralRocketLoader.load(GeneralRocketLoader.java:102)
at info.openrocket.swing.gui.util.OpenFileWorker.doInBackground(OpenFileWorker.java:73)
at info.openrocket.swing.gui.util.OpenFileWorker.doInBackground(OpenFileWorker.java:28)
at java.desktop/javax.swing.SwingWorker$1.call(SwingWorker.java:304)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.desktop/javax.swing.SwingWorker.run(SwingWorker.java:343)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
at java.base/java.lang.Thread.run(Thread.java:840)
```

--- a/core/src/main/java/info/openrocket/core/file/openrocket/importt/ComponentHandler.java
+++ b/core/src/main/java/info/openrocket/core/file/openrocket/importt/ComponentHandler.java
@@ -48,6 +48,15 @@ class ComponentHandler extends AbstractElementHandler {
 			throw Reflection.handleWrappedException(e);
 		}
 
+		// Malformed and legacy files may contain component hierarchies that the
+		// current model cannot represent. Ignore the complete subtree instead of
+		// aborting the document load in RocketComponent.addChild().
+		if (!parent.isCompatible(c)) {
+			warnings.add(Warning.fromString(c.getComponentName() + " cannot be attached to "
+					+ parent.getComponentName() + "; ignoring this component and its subcomponents."));
+			return null;
+		}
+
 		parent.addChild(c);
 
 		return new ComponentParameterHandler(c, context);

--- a/core/src/test/java/info/openrocket/core/file/openrocket/importt/ComponentHandlerTest.java
+++ b/core/src/test/java/info/openrocket/core/file/openrocket/importt/ComponentHandlerTest.java
@@ -1,0 +1,78 @@
+package info.openrocket.core.file.openrocket.importt;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+
+import org.junit.jupiter.api.Test;
+
+import info.openrocket.core.document.OpenRocketDocument;
+import info.openrocket.core.file.GeneralRocketLoader;
+import info.openrocket.core.file.RocketLoadException;
+import info.openrocket.core.rocketcomponent.AxialStage;
+import info.openrocket.core.rocketcomponent.BodyTube;
+import info.openrocket.core.rocketcomponent.PodSet;
+import info.openrocket.core.rocketcomponent.Rocket;
+import info.openrocket.core.util.BaseTestCase;
+
+/**
+ * Tests recovery from invalid component hierarchies in OpenRocket files.
+ */
+public class ComponentHandlerTest extends BaseTestCase {
+
+	/**
+	 * A legacy development version allowed pod sets directly below stages. The
+	 * invalid pod and its descendants should be ignored without losing valid
+	 * siblings that follow it.
+	 */
+	@Test
+	public void testIncompatibleComponentIsIgnoredWithWarning() throws RocketLoadException {
+		String xml = """
+				<?xml version="1.0" encoding="utf-8"?>
+				<openrocket version="1.10" creator="OpenRocket 15.03dev">
+				  <rocket>
+				    <name>Invalid component hierarchy</name>
+				    <subcomponents>
+				      <stage>
+				        <name>Sustainer</name>
+				        <subcomponents>
+				          <podset>
+				            <name>Invalid pod set</name>
+				            <subcomponents>
+				              <bodytube>
+				                <name>Ignored pod body</name>
+				              </bodytube>
+				            </subcomponents>
+				          </podset>
+				          <bodytube>
+				            <name>Valid body tube</name>
+				            <length>0.3</length>
+				            <thickness>0.001</thickness>
+				            <radius>0.02</radius>
+				          </bodytube>
+				        </subcomponents>
+				      </stage>
+				    </subcomponents>
+				  </rocket>
+				</openrocket>
+				""";
+
+		GeneralRocketLoader loader = new GeneralRocketLoader((File) null);
+		ByteArrayInputStream input = new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+		OpenRocketDocument document = loader.load(input, "invalid-component-hierarchy");
+
+		Rocket rocket = document.getRocket();
+		AxialStage stage = rocket.getStage(0);
+		assertEquals(1, stage.getChildCount());
+		BodyTube bodyTube = assertInstanceOf(BodyTube.class, stage.getChild(0));
+		assertEquals("Valid body tube", bodyTube.getName());
+		assertEquals(1, loader.getWarnings().size());
+		String expectedWarning = new PodSet().getComponentName() + " cannot be attached to "
+				+ stage.getComponentName() + "; ignoring this component and its subcomponents.";
+		assertEquals(expectedWarning,
+				loader.getWarnings().iterator().next().toString());
+	}
+}

